presto-main/src/main/java/com/facebook/presto/operator/MergeJoinOperator.java

```diff
@@ -0,0 +1,407 @@
+                rightPage = mergeJoinSource.nextPage();
+
+                if (rightPage == null && !mergeJoinSource.isFinishing()) {
+                    return null;
+                }
+
+                if (rightPage != null) {
+                    rightPosition = 0;
+                }
+                else {
+                    noMoreRightInput = true;
+                }
+                continue;
+            }
+
+            return null;
+        }
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing
+                && leftPage == null
+                && (pageBuilder == null || pageBuilder.isEmpty());
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public void close()
+    {
+        pageBuilder = null;
+        mergeJoinSource.close();
+    }
+
+    private Page doGetOutput()
+    {
+        // ran out of space during previous output
+        if (rightPagesIterator != null) {
+            Page output = joinAllLeftMatches();
+            if (output != null) {
+                return output;
+            }
+        }
+
+        // middle of right match
+        if (!mergeJoinRightPages.isEmpty() && !advanceRightToEndOfMatch()) {
+            // needs more input to find end of match
+            return null;
+        }
+
+        // no output in progress, but the leftPosition can match right side
+        if (leftPage != null && !mergeJoinRightPages.isEmpty()) {
+            Page output = joinAllLeftMatches();
+            if (output != null) {
+                return output;
+            }
+            else if (leftPage == null) {
+                return null;
+            }
+        }
+
+        // if at end, consume all remaining positions
+        if (leftPage == null || rightPage == null) {
+            if (leftPage != null && noMoreRightInput) {
+                return joinRemainingLeftPositions();
+            }
+            if (finishing && pageBuilder != null) {
+                return buildOutput();
+            }
+            return null;
+        }
+
+        // iterate more here to reduce number of returns
+        while (leftPage != null) {
+            prepareOutput();
+
+            // look for a new match starting with the leftPosition row on the left and rightPosition row on the right
+            int compareResult = compare();
+            while (compareResult < 0) {
+                if (advanceLeftToEndPosition()) {
+                    return null;
+                }
+                compareResult = compare();
+            }
+
+            while (compareResult > 0) {
+                if (advanceRightToEndPosition()) {
+                    return null;
+                }
+                compareResult = compare();
+            }
+
+            while (advanceNulls()) {
+                compareResult = compare();
+            }
+
+            if (compareResult == 0) {
+                // find all matching right positions
```

Comment:
Can we extract this logic to a helper function? what it does is "find the leftPosition and rightPosition for the next first matching row"